File: Sample-ObjC/07-add-n-delete-data/Mesosfer-ObjC/QueryDataViewController.m
/**
 * Copyright (c) 2016, Mesosfer
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "QueryDataViewController.h"
#import "UIViewController+Helper.h"
#import <Mesosfer/Mesosfer.h>
#import "FormDataViewController.h"

@interface QueryDataViewController () <FormDataViewControllerDelegate>

@property (nullable, nonatomic, strong) NSMutableArray<MFData *>* datas;

@end

@implementation QueryDataViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.tableView.estimatedRowHeight = 100;
    self.tableView.rowHeight = UITableViewAutomaticDimension;
    
    [self queryData];
}

- (void)queryData {
    NSString *bucket = @"Beacon";
    MFQuery *query = [MFData queryWithBucket:bucket];
    [query findAsyncWithBlock:^(NSArray * _Nullable objects, NSError * _Nullable error) {
        if (error) {
            [self showErrorWithTitle:@"Query data failed" error:error handler:nil];
            return;
        }
        
        if (objects) {
            self.title = [NSString stringWithFormat:@"Query '%@'", bucket];
            self.datas = [NSMutableArray arrayWithArray:objects];
            [self.tableView reloadData];
        }
    }];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.identifier isEqualToString:@"segueAdd"]) {
        FormDataViewController *formView = segue.destinationViewController;
        formView.delegate = self;
    }
}

#pragma mark - UITableViewDelegate

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    if (self.datas) {
        return self.datas.count;
    }
    
    return 0;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return 1;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    if (self.datas) {
        return self.datas[section].objectId;
    }
    return nil;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cellData" forIndexPath:indexPath];
    
    if (self.datas) {
        MFData *data = self.datas[indexPath.section];
        
        if ([data dictionary]) {
            cell.textLabel.text = [NSString stringWithFormat:@"%@", [data dictionary]];
        }
    }
    
    return cell;
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        if (self.datas) {
            MFData *data = self.datas[indexPath.section];
            [data deleteAsyncWithBlock:^(BOOL succeeded, NSError * _Nullable error) {
                if (succeeded) {
                    // remove deleted data from array
                    [self.datas removeObjectAtIndex:indexPath.section];
                    // update table data
                    [self.tableView beginUpdates];
                    [self.tableView deleteSections:[NSIndexSet indexSetWithIndex:indexPath.section]
                                  withRowAnimation:UITableViewRowAnimationAutomatic];
                    [self.tableView endUpdates];
                } else {
                    [self showErrorWithTitle:@"Delete data failed" error:error handler:nil];
                }
            }];
        }
    }
}

#pragma mark - FormDataViewControllerDelegate

- (void)onSavedDataBeacon:(MFData *)beacon isNewData:(BOOL)isNewData {
    if (isNewData) {
        // append new saved data into array
        if (!self.datas) {
            self.datas = [[NSMutableArray alloc] init];
        }
        [self.datas addObject:beacon];
        // update table data
        [self.tableView beginUpdates];
        [self.tableView insertSections:[NSIndexSet indexSetWithIndex:(self.datas.count - 1)]
                      withRowAnimation:UITableViewRowAnimationAutomatic];
        [self.tableView endUpdates];
    }
}

@end

File: Sample-Swift/07-add-n-delete-data/Mesosfer-Swift/QueryDataViewController.swift
//
//  QueryDataViewController.swift
//  Mesosfer-Swift
//
//  Created by Eyro Labs on 11/5/16.
//  Copyright © 2016 Eyro Labs. All rights reserved.
//

import Foundation
import UIKit
import Mesosfer

class QueryDataViewController: UITableViewController {
    var datas: [MFData]?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.tableView.estimatedRowHeight = 100
        self.tableView.rowHeight = UITableViewAutomaticDimension
        
        self.querydata()
    }
    
    func querydata() {
        let bucket = "Beacon"
        MFData.query(withBucket: bucket).findAsync { (datas, error) in
            if let e = error as? NSError {
                self.showError(title: "Query data failed", error: e)
                return
            }
            
            if let datas = datas as? [MFData] {
                self.title = "Query '\(bucket)'"
                self.datas = datas
                self.tableView.reloadData()
            }
        }
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == "segueAdd" {
            let formView = segue.destination as! FormDataViewController
            formView.delegate = self
        }
    }
}

extension QueryDataViewController {
    override func numberOfSections(in tableView: UITableView) -> Int {
        if let datas = self.datas {
            return datas.count
        }
        return 0
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 1
    }
    
    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        if let datas = self.datas {
            return datas[section].objectId
        }
        return nil
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "cellData", for: indexPath)
        
        if let datas = self.datas {
            let data = datas[indexPath.section]
            
            if let dict = data.dictionary() as? [String:AnyObject] {
                cell.textLabel?.text = "\(dict)"
            }
        }
        
        return cell
    }
    
    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
        if editingStyle == .delete {
            if let datas = self.datas {
                let data = datas[indexPath.section]
                data.deleteAsync(block: { (succeeded, error) in
                    if succeeded {
                        // remove deleted data from array
                        self.datas!.remove(at: indexPath.section)
                        // update table data
                        self.tableView.beginUpdates()
                        self.tableView.deleteSections(IndexSet(integer: indexPath.section), with: .automatic)
                        self.tableView.endUpdates()
                    } else {
                        self.showError(title: "Delete data failed", error: error as! NSError)
                    }
                })
            }
        }
    }
}

extension QueryDataViewController : FormDataViewControllerDelegate {
    func onSavedData(beacon: MFData, isNewData: Bool) {
        if isNewData {
            // append new saved data into array
            if self.datas == nil {
                self.datas = [MFData]()
            }
            self.datas!.append(beacon)
            // update table data
            self.tableView.beginUpdates()
            self.tableView.insertSections(IndexSet(integer: self.datas!.count - 1), with: .automatic)
            self.tableView.endUpdates()
        }
    }
}
